scylla-proxy/src/actions.rs

```diff
@@ -0,0 +1,135 @@
+use std::{sync::Arc, time::Duration};
+
+use bytes::Bytes;
+use rand::Rng;
+
+use crate::{
+    frame::{FrameOpcode, RequestOpcode, ResponseOpcode},
+    RequestFrame, ResponseFrame,
+};
+
+#[derive(Debug, Clone)]
+pub enum Condition {
+    True,
+    False,
+    Not(Box<Condition>),
+    And(Box<Condition>, Box<Condition>),
+    Or(Box<Condition>, Box<Condition>),
+    ConnectionSeqNo(usize),
+    RequestOpcode(RequestOpcode),
+    ResponseOpcode(ResponseOpcode),
+    BodyContainsCaseSensitive(Box<[u8]>),
+    BodyContainsCaseInsensitive(Box<[u8]>),
+    RandomWithProbability(f64),
+}
+
+pub(crate) struct EvaluationContext {
+    pub(crate) connection_seq_no: usize,
+    pub(crate) opcode: FrameOpcode,
+    pub(crate) frame_body: Bytes,
+}
+
+impl Condition {
+    pub(crate) fn eval(&self, ctx: &EvaluationContext) -> bool {
+        match self {
+            Condition::True => true,
+
+            Condition::False => false,
+
+            Condition::Not(c) => !c.eval(ctx),
+
+            Condition::And(c1, c2) => c1.eval(ctx) && c2.eval(ctx),
+
+            Condition::Or(c1, c2) => c1.eval(ctx) || c2.eval(ctx),
+
+            Condition::ConnectionSeqNo(no) => *no == ctx.connection_seq_no,
+
+            Condition::RequestOpcode(op1) => match ctx.opcode {
+                FrameOpcode::Request(op2) => *op1 == op2,
+                FrameOpcode::Response(_) => panic!(
+                    "Invalid type applied in rule condition: driver request opcode in cluster context"
+                ),
+            },
+
+            Condition::ResponseOpcode(op1) => match ctx.opcode {
+                FrameOpcode::Request(_) => panic!(
+                    "Invalid type applied in rule condition: cluster response opcode in driver context"
+                ),
+                FrameOpcode::Response(op2) => *op1 == op2,
+            },
+
+            Condition::BodyContainsCaseSensitive(pattern) => ctx
+                .frame_body
+                .windows(pattern.len())
+                .any(|window| *window == **pattern),
+
+            Condition::BodyContainsCaseInsensitive(pattern) => std::str::from_utf8(pattern)
+                .map(|pattern_str| {
+                    ctx.frame_body.windows(pattern.len()).any(|window| {
+                        std::str::from_utf8(window)
+                            .map(|window_str| unicase::eq_ascii(window_str, pattern_str))
+                            .unwrap_or(false)
+                    })
+                })
+                .unwrap_or(false),
```

Comment:
> I don't think it is guaranteed that an uppercase and lowercase variant of a Unicode character must have the same length when encoded as UTF-8.

Right, I should have mentioned this supports ASCII characters only.

> The frame may contain the string, but it may have different length.

What do you mean? Do you assume ASCII here or not?

Reply:
> > The frame may contain the string, but it may have different length.
> 
> What do you mean? Do you assume ASCII here or not?

I think I just didn't read it carefully enough and missed the `unicase::eq_ascii` (btw, we have [`eq_ignore_ascii_case`](https://doc.rust-lang.org/std/primitive.str.html#method.eq_ignore_ascii_case) in the standard library).